Add Check Support for Java 21 Pattern Matching for Switch Syntax: NeedBraces

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation: https://checkstyle.org/checks/blocks/needbraces.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="NeedBraces">
            <property name="tokens" value="LITERAL_CASE, LITERAL_DEFAULT"/>
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {

    void test(Object obj) {
        switch (obj) {
            case ColoredPoint(_, _, _):
            case Rectangle(_, _) : // violation
                break;
            default : // expected violation
        }
        switch (obj) {
            case ColoredPoint(int x, _, _) when (x >= 2)   // violation
                    :
                break;
            case ColoredPoint(int x, _, _)    // violation
                    when(x >= 2) :
                break;
            case ColoredPoint(int x, _, _)
                    when(x >= 2) : {}
            break;
            default :   // expected violation
        }
        switch (obj) {
            case ColoredPoint(int x, _, _) when (x >= 2) // violation
                    -> throw new RuntimeException();
            case ColoredPoint(int x, _, _)
                    when(x >= 2) -> {}
            default -> throw new RuntimeException();  // violation
        }
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:13: 'case' construct must use '{}'s. [NeedBraces]
[ERROR] D:\CS\test\src\Test.java:14:13: 'case' construct must use '{}'s. [NeedBraces]
[ERROR] D:\CS\test\src\Test.java:17:13: 'case' construct must use '{}'s. [NeedBraces]
[ERROR] D:\CS\test\src\Test.java:26:13: 'case' construct must use '{}'s. [NeedBraces]
[ERROR] D:\CS\test\src\Test.java:30:13: 'default' construct must use '{}'s. [NeedBraces]
Audit done.
Checkstyle ends with 5 errors.
PS D:\CS\test> 

```

---
Describe what you want in detail

we are ok here. This can be closed or add an input file for cases with `when`

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/NeedBracesCheckTest.java
@@ -276,4 +276,48 @@ public class NeedBracesCheckTest extends AbstractModuleTestSupport {
             expected);
     }
 
+    @Test
+    public void testPatternMatchingForSwitch() throws Exception {
+        final String[] expected = {
+            "17:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "21:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "25:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "28:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "53:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "55:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "58:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "61:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "74:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "78:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "80:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "83:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "90:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "91:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputNeedBracesPatternMatchingForSwitch.java"),
+            expected);
+    }
+
+    @Test
+    public void testPatternMatchingForSwitchAllowSingleLine() throws Exception {
+        final String[] expected = {
+            "17:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "21:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "25:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "28:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "55:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "58:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "61:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "74:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "80:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+            "83:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "default"),
+            "90:13: " + getCheckMessage(MSG_KEY_NEED_BRACES, "case"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputNeedBracesPatternMatchingForSwitchAllowSingleLine.java"),
+            expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/blocks/needbraces/InputNeedBracesPatternMatchingForSwitch.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/blocks/needbraces/InputNeedBracesPatternMatchingForSwitch.java
@@ -1,0 +1,101 @@
+/*
+NeedBraces
+allowSingleLineStatement = false
+allowEmptyLoopBody = (default)false
+tokens = LITERAL_CASE, LITERAL_DEFAULT, LAMBDA
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.blocks.needbraces;
+
+public class InputNeedBracesPatternMatchingForSwitch {
+
+    void testSwitchStatements(Object o) {
+        switch (o) {
+            case Integer i when (i == 0) :  // violation
+                System.out.println("zero");
+                break;
+
+            case String s when (s.equals("a")) :  // violation
+                System.out.println("a");
+                break;
+
+            case Point(int x, int y ) when (x>=0 && y >=0) :  // violation
+                System.out.println("point");
+                break;
+            default :  // violation
+                System.out.println("default");
+        }
+
+         switch (o) {
+            case Integer i when (i == 0) : {
+                System.out.println("zero");
+                break;
+            }
+            case String s when (s.equals("a")) : {
+                System.out.println("a");
+                break;
+            }
+            case Point(int x, int y ) when (x>=0 && y >=0) : {
+                System.out.println("point");
+                break;
+            }
+            default : {
+                System.out.println("default");
+            }
+        }
+    }
+
+    void testSwitchRule(Object o) {
+        switch (o) {
+            case Integer i when (i == 0) -> System.out.println("zero");   // violation
+
+            case String s when (s.equals("a")) ->  // violation
+                System.out.println("a");
+
+            case Point(int x, int y ) when (x>=0 && y >=0) ->  // violation
+                System.out.println("point");
+
+            default ->  // violation
+                System.out.println("default");
+        }
+        switch (o) {
+            case Integer i when (i == 0) -> {System.out.println("zero");}
+            case String s when (s.equals("a")) -> {System.out.println("a");}
+            case Point(int x, int y ) when (x>=0 && y >=0) -> {System.out.println("point");}
+            default -> {System.out.println("default");}
+        }
+    }
+
+    void testSwitchExpression(Object o) {
+        int a =  switch (o) {
+            case Integer i when (i == 0) :  // violation
+                System.out.println("zero");
+                yield i;
+
+            case String s when (s.equals("a")) : yield s.length();  // violation
+
+            case Point(int x, int y ) when (x>=0 && y >=0) :  // violation
+                yield x + y;
+
+            default :  // violation
+                yield 0;
+        };
+
+        int b =  switch (o) {
+            case Integer i when (i == 0) -> i;
+            // until https://github.com/checkstyle/checkstyle/issues/15322
+            case String s when (s.equals("a")) -> s.length();  // violation
+            case Point(int x, int y ) when (x>=0 && y >=0) ->  // violation
+                x + y;
+            case Integer i when (i == 9) -> {
+                int n = i;
+                yield n;
+            }
+            default -> 0;
+        };
+    }
+    record Point(int x, int y) {}
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/blocks/needbraces/InputNeedBracesPatternMatchingForSwitchAllowSingleLine.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/blocks/needbraces/InputNeedBracesPatternMatchingForSwitchAllowSingleLine.java
@@ -1,0 +1,100 @@
+/*
+NeedBraces
+allowSingleLineStatement = true
+allowEmptyLoopBody = (default)false
+tokens = LITERAL_CASE, LITERAL_DEFAULT, LAMBDA
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.blocks.needbraces;
+
+public class InputNeedBracesPatternMatchingForSwitchAllowSingleLine {
+
+    void testSwitchStatements(Object o) {
+        switch (o) {
+            case Integer i when (i == 0) :  // violation
+                System.out.println("zero");
+                break;
+
+            case String s when (s.equals("a")) :  // violation
+                System.out.println("a");
+                break;
+
+            case Point(int x, int y ) when (x>=0 && y >=0) :  // violation
+                System.out.println("point");
+                break;
+            default :  // violation
+                System.out.println("default");
+        }
+
+         switch (o) {
+            case Integer i when (i == 0) : {
+                System.out.println("zero");
+                break;
+            }
+            case String s when (s.equals("a")) : {
+                System.out.println("a");
+                break;
+            }
+            case Point(int x, int y ) when (x>=0 && y >=0) : {
+                System.out.println("point");
+                break;
+            }
+            default : {
+                System.out.println("default");
+            }
+        }
+    }
+
+    void testSwitchRule(Object o) {
+        switch (o) {
+            case Integer i when (i == 0) -> System.out.println("zero"); // ok, single line is true
+
+            case String s when (s.equals("a")) ->  // violation
+                System.out.println("a");
+
+            case Point(int x, int y ) when (x>=0 && y >=0) ->  // violation
+                System.out.println("point");
+
+            default ->  // violation
+                System.out.println("default");
+        }
+        switch (o) {
+            case Integer i when (i == 0) -> {System.out.println("zero");}
+            case String s when (s.equals("a")) -> {System.out.println("a");}
+            case Point(int x, int y ) when (x>=0 && y >=0) -> {System.out.println("point");}
+            default -> {System.out.println("default");}
+        }
+    }
+
+    void testSwitchExpression(Object o) {
+        int a =  switch (o) {
+            case Integer i when (i == 0) :  // violation
+                System.out.println("zero");
+                yield i;
+
+            case String s when (s.equals("a")) : yield s.length();  // ok, single line is true
+
+            case Point(int x, int y ) when (x>=0 && y >=0) :  // violation
+                yield x + y;
+
+            default :  // violation
+                yield 0;
+        };
+
+        int b =  switch (o) {
+            case Integer i when (i == 0) -> i;
+            case String s when (s.equals("a")) -> s.length();
+            case Point(int x, int y ) when (x>=0 && y >=0) ->  // violation
+                x + y;
+            case Integer i when (i == 9) -> {
+                int n = i;
+                yield n;
+            }
+            default -> 0;
+        };
+    }
+    record Point(int x, int y) {}
+}

